File: src/services/hextech.ts
// 海克斯大乱斗推荐：选英雄推荐 + 选海克斯牌推荐
import { getHextechHeroRank, getHextechRuneRank, getAugmentList, getHeroList } from '../api/cn101.js';
import { heroDisplayName, type AugmentInfo, type HextechHeroStat, type HextechRuneStat } from '../models.js';

export interface AugmentPick {
  augment: AugmentInfo;
  winRate: number;       // 持有胜率（百分比显示用 *100）
  pickRate: number;      // 出场率
  winRank: number;
  pickRank: number;
  bestHeroes: string[];  // 最适合英雄（中文名）
}

export interface HeroAugmentSuggestion {
  heroId: number;
  title: string;
  alias: string;
  winRate: number;       // 百分比
  pickRate: number;
  rank: number;
  bestAugments: (AugmentInfo & { winRate: number; pickRate: number; winRank: number })[];
  bestPartners: { heroId: number; title: string; alias: string; winRate: number }[];
}

/** 展示用胜率：接口返回 0.5773 表示 57.73% */
export const toPct = (v: number) => +(v * 100).toFixed(2);

/** 选牌推荐：全牌榜按胜率/出场率排序（带品质、适用英雄） */
export async function recommendAugments(topN = 20): Promise<AugmentPick[]> {
  const [runes, augments, heroes] = await Promise.all([getHextechRuneRank(), getAugmentList(), getHeroList()]);
  return runes
    .slice()
    .sort((a, b) => b.winRate - a.winRate)
    .slice(0, topN)
    .map((r: HextechRuneStat) => ({
      augment: augments.get(r.augmentId) ?? {
        augmentID: r.augmentId, name_cn: `牌#${r.augmentId}`, name_en: '', level: '',
        tooltip: '', small_Icon: '', large_Icon: '',
      },
      winRate: toPct(r.winRate),
      pickRate: toPct(r.pickRate),
      winRank: r.winRank,
      pickRank: r.pickRank,
      bestHeroes: r.bestHeroes.slice(0, 4).map((id) => heroDisplayName(heroes.get(id), id)),
    }));
}

/** 选英雄推荐：海克斯大乱斗英雄榜（胜率排序），附推荐海克斯牌与搭档 */
export async function recommendHextechHeroes(topN = 20): Promise<HeroAugmentSuggestion[]> {
  const [heroes, augments, list] = await Promise.all([
    getHeroList(),
    getAugmentList(),
    getHextechHeroRank(),
  ]);
  // 牌榜胜率映射（用于展示英雄推荐牌的胜率）
  const runes = await getHextechRuneRank().catch(() => []);
  const runeWin = new Map(runes.map((r) => [r.augmentId, toPct(r.winRate)]));
  const runePick = new Map(runes.map((r) => [r.augmentId, toPct(r.pickRate)]));
  const runeRank = new Map(runes.map((r) => [r.augmentId, r.winRank]));

  return list
    .slice()
    .sort((a: HextechHeroStat, b: HextechHeroStat) => b.winRate - a.winRate)
    .slice(0, topN)
    .map((h: HextechHeroStat) => ({
      heroId: h.heroId,
      title: heroDisplayName(heroes.get(h.heroId), h.heroId),
      alias: heroes.get(h.heroId)?.alias ?? '',
      winRate: toPct(h.winRate),
      pickRate: toPct(h.pickRate),
      rank: h.rank,
      bestAugments: h.bestAugments.slice(0, 3).map((id) => ({
        ...(augments.get(id) ?? { augmentID: id, name_cn: `牌#${id}`, name_en: '', level: '', tooltip: '', small_Icon: '', large_Icon: '' }),
        winRate: runeWin.get(id) ?? 0,
        pickRate: runePick.get(id) ?? 0,
        winRank: runeRank.get(id) ?? 0,
      })),
      // 最佳拍档：按组合胜率降序取前 3（官方字段按综合 rank 排序，与胜率顺序不一致）
      bestPartners: h.bestPartners
        .slice()
        .sort((a, b) => b.winRate - a.winRate)
        .slice(0, 3)
        .map((p) => ({
          heroId: p.heroId,
          title: heroDisplayName(heroes.get(p.heroId), p.heroId),
          alias: heroes.get(p.heroId)?.alias ?? '',
          winRate: toPct(p.winRate),
        })),
    }));
}

/** 最佳拍档榜：聚合官方英雄榜的搭档数据，按组合胜率排序（101 页面第三个内页签） */
export async function recommendHextechPartners(topN = 20): Promise<{
  heroId: number; heroTitle: string; heroAlias: string;
  partnerId: number; partnerTitle: string; partnerAlias: string;
  winRate: number; pickRate: number;
}[]> {
  const [heroes, list] = await Promise.all([getHeroList(), getHextechHeroRank()]);
  const rows: {
    heroId: number; heroTitle: string; heroAlias: string;
    partnerId: number; partnerTitle: string; partnerAlias: string;
    winRate: number; pickRate: number;
  }[] = [];
  for (const h of list) {
    for (const p of h.bestPartners) {
      rows.push({
        heroId: h.heroId,
        heroTitle: heroDisplayName(heroes.get(h.heroId), h.heroId),
        heroAlias: heroes.get(h.heroId)?.alias ?? '',
        partnerId: p.heroId,
        partnerTitle: heroDisplayName(heroes.get(p.heroId), p.heroId),
        partnerAlias: heroes.get(p.heroId)?.alias ?? '',
        winRate: toPct(p.winRate),
        pickRate: toPct(p.pickRate),
      });
    }
  }
  return rows.sort((a, b) => b.winRate - a.winRate).slice(0, topN);
}

/** 品质显示名 */
export const augmentLevelName = (l: string) =>
  ({ kSilver: '白银', kGold: '黄金', kPrismatic: '棱彩' }[l] ?? l);

File: desktop/src-tauri/src/hextech.rs
// 海克斯大乱斗推荐（对应 src/services/hextech.ts）：选英雄推荐 + 选海克斯牌推荐

use crate::cn101::cn;
use crate::models::*;
use std::collections::HashMap;

/// 展示用胜率：接口返回 0.5773 表示 57.73%
pub fn to_pct(v: f64) -> f64 {
    (v * 100.0 * 100.0).round() / 100.0
}

/// 品质显示名
pub fn augment_level_name(level: &str) -> String {
    match level {
        "kSilver" => "白银".to_string(),
        "kGold" => "黄金".to_string(),
        "kPrismatic" => "棱彩".to_string(),
        _ => level.to_string(),
    }
}

/// 选牌推荐：全牌榜按胜率排序（带品质、适用英雄）
pub async fn recommend_augments(top_n: i64) -> Result<Vec<AugmentPick>, String> {
    let (runes, augments, heroes) = tokio::join!(
        cn().get_hextech_rune_rank(None, false),
        cn().get_augment_list(false),
        cn().get_hero_list(false),
    );
    let (runes, augments, heroes) = (runes?, augments?, heroes?);
    let mut sorted = runes.clone();
    sorted.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));
    Ok(sorted.into_iter().take(top_n.max(0) as usize).map(|r| AugmentPick {
        augment: augments.get(&r.augment_id).cloned().unwrap_or_else(|| AugmentInfo::fallback(r.augment_id)),
        win_rate: to_pct(r.win_rate),
        pick_rate: to_pct(r.pick_rate),
        win_rank: r.win_rank,
        pick_rank: r.pick_rank,
        best_heroes: r.best_heroes.iter().take(4).map(|id| hero_display_name(heroes.get(id), *id)).collect(),
    }).collect())
}

/// 选英雄推荐：海克斯大乱斗英雄榜（胜率排序），附推荐海克斯牌与搭档
pub async fn recommend_hextech_heroes(top_n: i64) -> Result<Vec<HeroAugmentSuggestion>, String> {
    let (heroes, augments, list, runes) = tokio::join!(
        cn().get_hero_list(false),
        cn().get_augment_list(false),
        cn().get_hextech_hero_rank(None, false),
        cn().get_hextech_rune_rank(None, false),
    );
    let (heroes, augments, list, runes) = (heroes?, augments?, list?, runes.ok().unwrap_or_default());
    let rune_win: HashMap<i64, f64> = runes.iter().map(|r| (r.augment_id, to_pct(r.win_rate))).collect();
    let rune_pick: HashMap<i64, f64> = runes.iter().map(|r| (r.augment_id, to_pct(r.pick_rate))).collect();
    let rune_rank: HashMap<i64, i64> = runes.iter().map(|r| (r.augment_id, r.win_rank)).collect();

    let mut sorted = list.clone();
    sorted.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));
    Ok(sorted.into_iter().take(top_n.max(0) as usize).map(|h| {
        let best_augments = h.best_augments.iter().take(3).map(|id| {
            let a = augments.get(id).cloned().unwrap_or_else(|| AugmentInfo::fallback(*id));
            BestAugment {
                name_cn: a.name_cn,
                win_rate: rune_win.get(id).copied().unwrap_or(0.0),
                pick_rate: rune_pick.get(id).copied().unwrap_or(0.0),
                win_rank: rune_rank.get(id).copied().unwrap_or(0),
            }
        }).collect();
        let mut partners: Vec<&PartnerRow> = h.best_partners.iter().collect();
        partners.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));
        let best_partners = partners.into_iter().take(3).map(|p| BestPartner {
            hero_id: p.hero_id,
            title: hero_display_name(heroes.get(&p.hero_id), p.hero_id),
            alias: heroes.get(&p.hero_id).map(|h| h.alias.clone()).unwrap_or_default(),
            win_rate: to_pct(p.win_rate),
        }).collect();
        HeroAugmentSuggestion {
            hero_id: h.hero_id,
            title: hero_display_name(heroes.get(&h.hero_id), h.hero_id),
            alias: heroes.get(&h.hero_id).map(|h| h.alias.clone()).unwrap_or_default(),
            win_rate: to_pct(h.win_rate),
            pick_rate: to_pct(h.pick_rate),
            rank: h.rank,
            best_augments,
            best_partners,
        }
    }).collect())
}

/// 最佳拍档榜：聚合官方英雄榜的搭档数据，按组合胜率排序
pub async fn recommend_hextech_partners(top_n: i64) -> Result<Vec<PartnerRankRow>, String> {
    let (heroes, list) = tokio::join!(
        cn().get_hero_list(false),
        cn().get_hextech_hero_rank(None, false),
    );
    let (heroes, list) = (heroes?, list?);
    let mut rows: Vec<PartnerRankRow> = Vec::new();
    for h in &list {
        for p in &h.best_partners {
            rows.push(PartnerRankRow {
                hero_id: h.hero_id,
                hero_title: hero_display_name(heroes.get(&h.hero_id), h.hero_id),
                hero_alias: heroes.get(&h.hero_id).map(|x| x.alias.clone()).unwrap_or_default(),
                partner_id: p.hero_id,
                partner_title: hero_display_name(heroes.get(&p.hero_id), p.hero_id),
                partner_alias: heroes.get(&p.hero_id).map(|x| x.alias.clone()).unwrap_or_default(),
                win_rate: to_pct(p.win_rate),
                pick_rate: to_pct(p.pick_rate),
            });
        }
    }
    rows.sort_by(|a, b| b.win_rate.partial_cmp(&a.win_rate).unwrap_or(std::cmp::Ordering::Equal));
    rows.truncate(top_n.max(0) as usize);
    Ok(rows)
}
